util: fixed memory leaks in zenfs utility

percona-ysorokin opened the pull request:
Fixed leak in 'ZenFS::MkFS()' caused by a 'Superblock' allocated on
heap and never deallocated.

Fixed resource leak in 'ListZenFileSystems()' where 'opendir()' did not
have corresponding 'closedir()'

Reworked 'zbd_open()' and 'zenfs_mount()' functions so that they now
accept/return 'std::unique_ptr's for 'ZonedBlockDevice' and 'ZenFS'
instead of raw pointers. This fixed a number of memory leaks in
'zenfs' utility command handlers.

Fixed 'std::string's passed by value instead of by const reference in
some function signatures.

Signed-off-by: Yura Sorokin <yura.sorokin@percona.com>

yhr commented on util/zenfs.cc:
```diff
@@ -36,36 +37,39 @@ DEFINE_string(backup_path, "", "Path to backup files");
 
-ZonedBlockDevice *zbd_open(bool readonly, bool exclusive) {
-  ZonedBlockDevice *zbd = new ZonedBlockDevice(FLAGS_zbd, nullptr);
+using ZonedBlockDevicePtr = std::unique_ptr<ZonedBlockDevice>;
```
Please don't introduce new types unless it is really needed, we want to stick with rocksdb coding style (and I think adding types just makes the code harder to read - just another level of indirection)

yhr commented on util/zenfs.cc:
```diff
@@ -79,41 +83,40 @@ int zenfs_tool_mkfs() {
 
-  ZonedBlockDevice *zbd = zbd_open(false, true);
-  if (zbd == nullptr) return 1;
+  auto zbd = zbd_open(false, true);
```
The only place I want auto to be used is for iterators in for loops - I'm very much in favor of typed languages :)

yhr commented on util/zenfs.cc:
```diff
@@ -79,41 +83,40 @@ int zenfs_tool_mkfs() {
 
   zbd = zbd_open(false, true);
-  zenFS = new ZenFS(zbd, FileSystem::Default(), nullptr);
+  auto *zbdRaw = zbd.get();
```
See prev comment

yhr commented on util/zenfs.cc:
```diff
@@ -167,12 +169,11 @@ void format_path(std::string &path)
-  ZenFS *zenFS;
-  s = zenfs_mount(zbd, &zenFS, true);
+  ZenFSPtr zenFS;
+  auto s = zenfs_mount(zbd, &zenFS, true);
```
Same here, no need to use auto

yhr commented on util/zenfs.cc:
```diff
@@ -186,20 +187,20 @@ int zenfs_tool_list() {
-  ZenFS *zenFS;
-  s = zenfs_mount(zbd, &zenFS, true);
+  ZenFSPtr zenFS;
+  auto s = zenfs_mount(zbd, &zenFS, true);
```
Please use full type declaration

yhr commented on util/zenfs.cc:
```diff
@@ -212,18 +213,17 @@ int zenfs_tool_df() {
 
-  for (it = zenFileSystems.begin(); it != zenFileSystems.end(); it++)
-    fprintf(stdout, "%s\t%s\n", it->first.c_str(), it->second.c_str());
+  for (const auto &p: zenFileSystems)
```
here auto makes sense

yhr commented on util/zenfs.cc:
```diff
@@ -274,56 +274,52 @@ IOStatus zenfs_tool_copy_file(FileSystem *f_fs, std::string f, FileSystem *t_fs,
-  if (buffer == nullptr) {
-    return IOStatus::IOError("Failed to allocate copy buffer");
-  }
+  buffer.resize(buffer_sz);
```
It's nice to reduce the number of lines, but an allocation failure can still happen, right?

yhr commented on util/zenfs.cc:
```diff
@@ -356,36 +352,34 @@ IOStatus zenfs_tool_copy_dir(FileSystem *f_fs, std::string f_dir, FileSystem *t_
   IOStatus io_status;
-  ZonedBlockDevice *zbd;
-  ZenFS *zenFS;
+  auto zbd = zbd_open(true, true);
```
No auto declarations please

yhr commented on util/zenfs.cc:
```diff
@@ -410,17 +401,18 @@ int zenfs_tool_restore() {
 
-  zbd = zbd_open(false, true);
-  if (zbd == nullptr) return 1;
+  auto zbd = zbd_open(false, true);
```
No autos

yhr commented on util/zenfs.cc:
```diff
@@ -431,12 +423,12 @@ int zenfs_tool_restore() {
-  s = zenfs_mount(zbd, &zenFS, true);
+  ZenFSPtr zenFS;
+  auto s = zenfs_mount(zbd, &zenFS, true);
   if (!s.ok()) {
```
Same here, please write out the type in stead

yhr reviewed (changes requested):
Thanks for the PR, looks good over all - but there are some nitpicks (mostly coding style) that needs to be addressed.

It would have been easier to review if the different fixes had been split up in individual patches, please try to do that next time. Small, one logical change per commit, patches helps when bisecting for regressions as well.

yhr commented:
zenfs integration testing ((util, smoke, quick performance test sets) passed

percona-ysorokin commented:
@yhr I addressed your comments:
- removed `ZonedBlockDevicePtr` and `ZenFSPtr` typedefs.
- `auto`s changed to explicit types
- removed formatting fixes
- reimplemented OOM check after buffer allocation in `zenfs_tool_copy_file()`
- removed utility help line fix (will be submitted as a separate PR)

yhr commented:
Looks good to me and passed integration testing
